iree/task/poller.c

```diff
@@ -0,0 +1,524 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/task/poller.h"
+
+#include "iree/base/tracing.h"
+#include "iree/task/executor.h"
+#include "iree/task/executor_impl.h"
+#include "iree/task/submission.h"
+#include "iree/task/task_impl.h"
+#include "iree/task/tuning.h"
+
+static int iree_task_poller_main(iree_task_poller_t* poller);
+
+iree_status_t iree_task_poller_initialize(
+    iree_task_executor_t* executor,
+    iree_thread_affinity_t ideal_thread_affinity,
+    iree_task_poller_t* out_poller) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  out_poller->executor = executor;
+  out_poller->ideal_thread_affinity = ideal_thread_affinity;
+  iree_notification_initialize(&out_poller->state_notification);
+  iree_atomic_task_slist_initialize(&out_poller->mailbox_slist);
+  iree_task_list_initialize(&out_poller->wait_list);
+
+  iree_task_poller_state_t initial_state = IREE_TASK_POLLER_STATE_RUNNING;
+  // TODO(benvanik): support initially suspended wait threads. This can reduce
+  // startup time as we won't give the system a chance to deschedule the calling
+  // thread as it performs the initial resume of the wait thread. We'll need to
+  // check in enqueue to see if the wait thread needs to be resumed.
+  // initial_state = IREE_TASK_POLLER_STATE_SUSPENDED;
+  iree_atomic_store_int32(&out_poller->state, initial_state,
+                          iree_memory_order_seq_cst);
+
+  // Acquire an event we can use to wake the wait thread from other threads.
+  iree_status_t status = iree_event_pool_acquire(
+      iree_task_executor_event_pool(out_poller->executor), 1,
+      &out_poller->wake_event);
+
+  // Wait set used to batch syscalls for polling/waiting on wait handles.
+  // This is currently limited to a relatively small max to make bad behavior
+  // clearer with nice RESOURCE_EXHAUSTED errors. If we start to hit that limit
+  // (~63+ simultaneous system waits) we'll need to shard out the wait sets -
+  // possibly with multiple wait threads (one per set).
+  if (iree_status_is_ok(status)) {
+    status = iree_wait_set_allocate(IREE_TASK_EXECUTOR_MAX_OUTSTANDING_WAITS,
+                                    executor->allocator, &out_poller->wait_set);
+  }
+  if (iree_status_is_ok(status)) {
+    status = iree_wait_set_insert(out_poller->wait_set, out_poller->wake_event);
+  }
+
+  iree_thread_create_params_t thread_params;
+  memset(&thread_params, 0, sizeof(thread_params));
+  thread_params.name = iree_make_cstring_view("iree-poller");
+  thread_params.create_suspended = false;
+  // TODO(benvanik): make high so to reduce latency? The sooner we wake the
+  // sooner we get ready tasks back in the execution queue, though we don't
+  // want to preempt any of the workers.
```

Comment:
Think any changes in here would show up on our benchmarks? If not, what sorts of workloads and/or devices would be most affected? Programs with high parallelism (lots of concurrent tasks)? Devices with lots of cores?

Reply:
Not today as we aren't using waits yet (they were pretty broken here!), but as soon as we have semaphores chaining submissions we'll hit this path. Having some generic benchmarks that test things like submitting N chained command buffers with semaphores and such would be useful, as would some more targeted microbenchmarks here. They aren't easy numbers to track over time or across devices, though, as there's so much platform influence.